mediaapi/routing/preview_url.go

```diff
@@ -0,0 +1,400 @@
+	"github.com/matrix-org/dendrite/mediaapi/types"
+	"github.com/matrix-org/dendrite/setup/config"
+	"github.com/matrix-org/util"
+	log "github.com/sirupsen/logrus"
+)
+
+// the parameters included in the incoming preview_url request
+type PreviewUrlRequest struct {
+	url url.URL
+	ts  types.UnixMs
+}
+
+// metadata of the url (media/html) to be previewed
+type mediaInfo struct {
+	MediaMetadata *types.MediaMetadata
+	Logger        *log.Entry
+}
+
+// PreviewUrlResponse defines the format of the JSON response
+// https://matrix.org/docs/spec/client_server/latest#get-matrix-media-r0-preview-url
+// TODO: add more fields to the response
+type PreviewUrlResponse struct {
+	MatrixImageSize int64  `json:"matrix:image:size"`
+	OgImage         string `json:"og:image"`
+	OgSiteName      string `json:"og:site_name"`
+	OgType          string `json:"og:type"`
+	OgTitle         string `json:"og:title"`
+	OgUrl           string `json:"og:url"`
+	OgDescription   string `json:"og:description"`
+}
+
+// PreviewUrl implements GET /preview_url.
+// Current implementation gets the url, parses the meta tags to obtain OG
+// data and returns a JSONResponse for the client to process.
+// TODO: avoid this endpoint in encrypted rooms
+func PreviewUrl(
+	req *http.Request, cfg *config.MediaAPI, db storage.Database) util.JSONResponse {
+
+	// parsing the request
+	r, err := parseRequest(req)
+	if err != nil {
+		return *err
+	}
+
+	// downloading metadata for the url
+	mediaInfo, err := r.downloadUrl(cfg, req.Context(), db)
+	if err != nil {
+		return *err
+	}
+
+	// preparing the response
+	response, err := mediaInfo.prepareResponse(cfg)
+	if err != nil {
+		return *err
+	}
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: response,
+	}
+}
+
+// parseRequest parses the incoming preview request to extract the url and ts.
+// Returns either a parsed PreviewUrlRequest or error formatted as
+// util.JSONResponse
+func parseRequest(req *http.Request) (*PreviewUrlRequest, *util.JSONResponse) {
+
+	// get the url to be previewed from the request
+	urlToPreview := req.URL.Query().Get("url")
+	if len(urlToPreview) == 0 {
+		return nil, &util.JSONResponse{
+			Code: http.StatusBadRequest,
+			JSON: jsonerror.MissingArgument("Missing URL"),
+		}
+	}
+
+	// parse the url
+	parsedUrl, err := url.Parse(urlToPreview)
+	if err != nil {
+		return nil, &util.JSONResponse{
+			Code: http.StatusBadRequest,
+			JSON: jsonerror.InvalidArgumentValue(
+				"Unable to parse url " + err.Error()),
+		}
+	}
+
+	// get the ts if provided in the request
+	if tsStr := req.URL.Query().Get("ts"); len(tsStr) > 0 {
+		ts, err := strconv.ParseInt(tsStr, 10, 64)
+		if err != nil {
+			return nil, &util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.InvalidArgumentValue(
+					"Couldn't parse 'ts' to a valid integer" + err.Error()),
+			}
+		}
+
+		request := &PreviewUrlRequest{
+			url: *parsedUrl,
+			// Convert timestamp to ms
+			ts: types.UnixMs(ts / 1000000),
```

Comment:
`tsStr` will be in milliseconds already looking at https://github.com/matrix-org/synapse/blob/4b965c862dc66c0da5d3240add70e9b5f0aa720b/synapse/rest/media/v1/preview_url_resource.py#L198 so it should just be fine to do `ts: types.UnixMs(ts)`.